src/host/AccessibilityNotifier.cpp

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "AccessibilityNotifier.h"
+
+#include "../interactivity/inc/ServiceLocator.hpp"
+
+using namespace Microsoft::Console;
+using namespace Microsoft::Console::Interactivity;
+
+template<typename U, typename T>
+constexpr U castSaturated(T v) noexcept
+{
+    static_assert(sizeof(U) <= sizeof(T), "use this for narrowing");
+    constexpr T min = std::numeric_limits<U>::min();
+    constexpr T max = std::numeric_limits<U>::max();
+    return gsl::narrow_cast<U>(v < min ? min : (v > max ? max : v));
+}
+
+AccessibilityNotifier::AccessibilityNotifier()
+{
+    // Mirrors _timerEmitMSAA / _timerEmitUIA
+    memset(&_state, 0, sizeof(_state));
+}
+
+AccessibilityNotifier::~AccessibilityNotifier()
+{
+    SetUIAProvider(nullptr);
+}
+
+void AccessibilityNotifier::Initialize(HWND hwnd, DWORD msaaDelay, DWORD uiaDelay) noexcept
+{
+    _hwnd = hwnd;
+
+    // delay=INFINITE is intended to disable events completely, but realistically,
+    // even a delay of 1s makes little sense. So, the cut-off was set to 10s.
+    if (msaaDelay < 10000 && hwnd)
+    {
+        _msaaEnabled = true;
+
+        // msaaDelay=0 makes all events synchronous. That's how
+        // it used to work and has a huge performance impact.
+        if (msaaDelay != 0)
+        {
+            // Convert from milliseconds to 100-nanosecond intervals.
+            // Negative values indicate relative time.
+            _msaaDelay = static_cast<int64_t>(msaaDelay) * -10000;
+        }
+    }
+
+    if (uiaDelay < 10000)
+    {
+        _uiaEnabled = true;
+
+        if (uiaDelay != 0)
+        {
+            _uiaDelay = static_cast<int64_t>(uiaDelay) * -10000;
+        }
+    }
+
+    if (_msaaDelay || _uiaDelay)
+    {
+        _timer.reset(_createTimer(&_timerEmitMSAA));
+    }
+
+    // Triggers the computation of _delay and _delayWindow.
+    SetUIAProvider(nullptr);
+}
+
+void AccessibilityNotifier::SetUIAProvider(IRawElementProviderSimple* provider) noexcept
+{
+    // NOTE: The assumption is that you're holding the console lock when calling any of the member functions.
+    // This is why we can safely update these members (no worker thread is running nor can be scheduled).
+    assert(ServiceLocator::LocateGlobals().getConsoleInformation().IsConsoleLocked());
+
+    // If UIA events are disabled, don't set _uiaProvider either.
+    // It would trigger unnecessary work.
+    if (!_uiaEnabled)
+    {
+        return;
+    }
+
+    // Of course we must ensure our precious provider object doesn't go away.
+    if (provider)
+    {
+        provider->AddRef();

```

Comment:
can we not store it in a `com_ptr`? is it because we doin' manual atomic stuff with it?